libvips/iofuncs/threadset.c

```diff
@@ -0,0 +1,330 @@
+                 */
+                if( vips__thread_profile ) 
+                        vips__thread_profile_attach( member->domain );
+
+                /* Execute the task.
+                 */
+                member->func( member->data, member->user_data );
+
+                /* Free any thread-private resources -- they will not be
+                 * useful for the next task to use this thread.
+                 */
+                vips_thread_shutdown();
+
+                member->domain = NULL;
+                member->func = NULL;
+                member->data = NULL;
+                member->user_data = NULL;
+
+                /* We are free ... back on the free list!
+                 */
+                g_mutex_lock( set->lock );
+                set->free = g_slist_prepend( set->free, member );
+                g_mutex_unlock( set->lock );
+        }
+
+        /* Kill has been requested. We leave this thread on the members 
+         * list so it can be found and joined.
+         */
+
+        return( NULL );
+}
+
+/* Create a new idle member for the set.
+ */
+static VipsThreadsetMember *
+vips_threadset_add( VipsThreadset *set )
+{
+        VipsThreadsetMember *member;
+
+        if( set->max_threads &&
+                set->n_threads >= set->max_threads ) {
+                vips_error( "VipsThreadset", 
+                        "%s", _( "threadset is exhausted" ) );
+                return( NULL );
+        }
+
+        member = g_new0( VipsThreadsetMember, 1 );
+        member->set = set;
+
+	vips_semaphore_init( &member->idle, 0, "idle" );
+
+	if( !(member->thread = vips_g_thread_new( "libvips worker", 
+                vips_threadset_work, member )) ) {
+                vips_semaphore_destroy( &member->idle );
+                VIPS_FREE( member );
+
+                return( NULL );
+	}
+
+        g_mutex_lock( set->lock );
+        set->members = g_slist_prepend( set->members, member );
+        set->n_threads += 1;
+        set->n_threads_highwater = 
+                VIPS_MAX( set->n_threads_highwater, set->n_threads );;
+        g_mutex_unlock( set->lock );
+
+        return( member );
+}
+
+/** 
+ * vips_threadset_new:
+ * @max_threads: maxium number of system threads
+ *
+ * Create a new threadset. 
+ *
+ * If @max_threads is 0, new threads will be created when necessary by
+ * vips_threadset_run(), with no limit on the number of threads.
+ *
+ * If @max_threads is > 0, then that many threads will be created by
+ * vips_threadset_new() during startup and vips_threadset_run() will fail if
+ * no free threads are available.
+ *
+ * Returns: the new threadset.
+ */
+VipsThreadset *
+vips_threadset_new( int max_threads )
+{
+        VipsThreadset *set;
+
+        set = g_new0( VipsThreadset, 1 );
+	set->lock = vips_g_mutex_new();
+	set->max_threads = max_threads;
+
+        if( set->max_threads > 0 )
+                for( int i = 0; i < set->max_threads; i++ ) {
```

Comment:
Should we declare `i` at the beginning of the scope block to ensure C89 compat?

Reply:
I thought we decided to switch to C99? I'll see if I can find the discussion.

Reply:
We build with meson in C gnu99 mode, in line with glib, so I think this means we are effectively a c99 project now.